### DISCUSSION TITLE
kubernetes service discovery of Apisix works or not?

### DISCUSSION THREAD

**Discussion by tirelibirefe** [apache/apisix#6782] [Q&A]
Hello,

How can I check Kubernetes service discovery of Apisix works or not?

Thanks & Regards

**Comment by tokers** [comment]
@tirelibirefe See https://apisix.apache.org/docs/apisix/next/discovery/kubernetes/

After you configure it, create Upstream to use the Kubernetes discovery mechanism and create a route to access this upstream to verify if it's in effective.

**Comment by tirelibirefe** [comment]
Hello @tokers 
Thanks for the answer.

I defined "the upstream". 
Kibana service is a "https" backend.

```
timeout:
  connect: 6
  send: 6
  read: 6
type: roundrobin
checks:
  active:
    concurrency: 10
    healthy:
      http_statuses:
        - 200
        - 302
      interval: 1
      successes: 2
    host: kibana-tracing-kb-http.es-tracing.svc
    http_path: /
    https_verify_certificate: true
    port: 80
    timeout: 1
    type: https
    unhealthy:
      http_failures: 5
      http_statuses:
        - 429
        - 404
        - 500
        - 501
        - 502
        - 503
        - 504
        - 505
      interval: 1
      tcp_failures: 2
      timeouts: 3
scheme: https
discovery_type: dns
pass_host: pass
name: kibana-es-tracing
service_name: kibana-tracing-kb-http.es-tracing.svc
tls:
  client_cert: |-
    MIIEiTCCAvGgAwIBAgIRALeALrZPDkFF4pnLSx85IGgwDQYJKoZIhvcNAQELBQAw
  client_key: |-
    MIIEpAIBAAKCAQEAoGlvIzkDr1vaz6lMlv45RJ3C8PGUr+xAU9ONB8rxqQqH+Rfz
i1QMiO/Rnchhi7r4
    B8PHdraoACG364BsikloTntNQSuNZTIBcnxIqAZi7oSKbd24zViOw5Y1HsXycT+6
keepalive_pool:
  idle_timeout: 60
  requests: 1000
  size: 320
```

here is the service of Kibana:

```
NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
apm-server-tracing-apm-http   ClusterIP   172.20.66.170   <none>        8200/TCP   3d12h
kibana-tracing-kb-http        ClusterIP   172.20.224.57   <none>        5601/TCP   9d
tracing-es-default            ClusterIP   None            <none>        9200/TCP   9d
tracing-es-http               ClusterIP   172.20.11.21    <none>        9200/TCP   9d
tracing-es-internal-http      ClusterIP   172.20.123.78   <none>        9200/TCP   9d
tracing-es-transport          ClusterIP   None            <none>        9300/TCP   9d
```

... then I defined "the route" and bounded to the upstream:

Upstream Type: ServiceDiscovery
Discovery Type: DNS

```
uri: /*
name: kibana-es-tracing
host: kibana-tracing.myawesomedomain.com
upstream_id: '401373256600257120'
enable_websocket: true
status: 1
```
This is important:
`upstream_id: '401373256600257120'`

When I try to connect from browser, I get  the following error messages:

```
2022/04/02 22:38:40 [error] 45#45: *8352188 [lua] init.lua:284: get_upstream_by_id(): failed to find upstream by id: 401373256600257120, client: 10.61.34.0, server: _, request: "GET / HTTP/1.1", host: "kibana-tracing.myawesomedomain.com"
2022/04/02 22:38:40 [error] 44#44: *8350339 [lua] init.lua:284: get_upstream_by_id(): failed to find upstream by id: 401373256600257120, client: 10.60.20.7, server: _, request: "GET /favicon.ico HTTP/1.1", host: "kibana-tracing.myawesomedomain.com", referrer: "https://kibana-tracing.myawesomedomain.com/"
10.60.20.7 - - [02/Apr/2022:22:38:40 +0000] kibana-tracing.myawesomedomain.com "GET /favicon.ico HTTP/1.1" 502 556 0.000 "https://kibana-tracing.myawesomedomain.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36" - - - "http://kibana-tracing.myawesomedomain.com"
10.61.34.0 - - [02/Apr/2022:22:38:40 +0000] kibana-tracing.myawesomedomain.com "GET / HTTP/1.1" 502 556 0.000 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36" - - - "http://kibana-tracing.myawesomedomain.com"
```

Does Apisix not discover the service or something else I missed? Am I doing something wrong or did I miss something? Could you please advise?

Thanks & Regards

**Reply by tokers** [reply]
I have several doubts.

1. How did you deploy APISIX? In which way you configure these routes and upstreams?
2. The title is for kubernetes service discovery, but why you use the dns way?
3. How did you configure the DNS service discovery in APISIX config.yaml?

**Reply by tirelibirefe** [reply]
I agree. DNS or anything like that mustn't be needed on K8s environment. Actually I didn't understand the purpose of service discovery of Apisix on K8s environment, it was seem very very primitive to me. Also when I noticed Apisix doesn't support https backends like other normal ingress controllers, I replaced Apisix from Kubernetes with a normal ingress controller which supports https backend.

Thanks for your advise and message.

**Reply by a11enhuang** [reply]
Is there any difference between directly using hostname in kubernetes environment and using kubernetes discovery? Because if APISIX is deployed in kubernetes, the service can be accessed directly through the host.
